modules/spark/src/main/scala/org/apache/ignite/spark/impl/package.scala

```diff
@@ -124,10 +127,41 @@ package object impl {
     def sqlTableInfo[K, V](ignite: Ignite, tabName: String,
         schemaName: Option[String]): Option[(CacheConfiguration[K, V], QueryEntity)] =
         cachesForSchema[K,V](ignite, schemaName)
-            .map(ccfg ⇒ ccfg.getQueryEntities.find(_.getTableName.equalsIgnoreCase(tabName)).map(qe ⇒ (ccfg, qe)))
+            .map(ccfg ⇒ refreshQueryEntity(ignite, ccfg,
+                ccfg.getQueryEntities.find(_.getTableName.equalsIgnoreCase(tabName))).map(qe ⇒ (ccfg, qe)))
             .find(_.isDefined)
             .flatten
 
+    /**
+      * @param ignite Ignite instance.
+      * @param ccfg Cache Configuration.
+      * @param qe Query Entity.
+      * @tparam K Key class.
+      * @tparam V Value class.
+      * @return QueryEntity for a given table.
+      */
+    def refreshQueryEntity[K, V](ignite: Ignite, ccfg : CacheConfiguration[K, V], qe: Option[QueryEntity]): Option[QueryEntity] =
+        if (qe.isDefined) Some(qe.get.setFields(refreshFields(ignite, ccfg, qe.get.getTableName)))
+        else qe
+
+    /**
+      * @param ignite Ignite instance.
+      * @param ccfg Cache Configuration.
+      * @param tabName Table name.
+      * @tparam K Key class.
+      * @tparam V Value class.
+      * @return HashMap consists of fields and fields type for a given table.
+      */
+	def refreshFields[K, V](ignite: Ignite, ccfg: CacheConfiguration[K, V], tabName: String): util.LinkedHashMap[String, String] = {
```

Comment:
We shouldn't modify `QueryEntity` content.
Seems we can completely replace usages of `QueryEntity` with `GridQueryTypeDescriptor`.
Let's do it

What do you think?

Reply:
Ok, I'll implement using GridQueryTypeDescriptor to replace QueryEntity.

One question here, do we still need CacheConfiguration?
Seems the only usage for CacheConfiguration is to get schema from it, but now GridQueryTypeDescriptor has schema field.

Reply:
> do we still need CacheConfiguration?

It seems that we don't.
If it really not necessary, we can remove it from this code.

Reply:
During implantation, I found out that GridQueryTypeDescriptor doesn't have keyfields set like QueryEntity.
The only way to get key fields is to iterate props map of QueryTypeDescriptorImpl, based on value of GridQueryProperty.key to determine if that field is a key field.
And this approach does not work with composite object key field if the table is created by cache API.

Does this approach to get key fields makes sense to you?

Reply:
@ldzhjn

Seems, we can iterate over GridQueryTypeDescriptor#fields and call GridQueryTypeDescriptor#property -> GridQueryProperty#key to get all key fields.

This calls doesn't require to use implementation specific methods, all methods are belong to interfaces.

Does it makes sense for you?

Reply:
@ldzhjn

Seems, we can iterate over GridQueryTypeDescriptor#fields and call GridQueryTypeDescriptor#property -> GridQueryProperty#key to get all key fields.

This calls doesn't require to use implementation specific methods, all methods are belong to interfaces.

Does it makes sense for you?

Reply:
Thanks for the heads up, I'll use this way then.

One more question here since you mentioned not to use implementation specific methods.
In line 123 of IgniteSQLRelation, we need to get alias for field name.
Currently my implementation is 
name = table.asInstanceOf[QueryTypeDescriptorImpl].aliases.getOrDefault(name, name),

Do I have to change the implementation here?

Reply:
Currently, we doesn't have another option to get aliases, so I propose to do following:

1. Use `table.asInstanceOf[QueryTypeDescriptorImpl]`
2. Fill the ticket to bring `aliases` to `GridQueryTypeDescriptor` interface.

Reply:
I have implemented the fix, please review and comment.
The tests in teamcity is all green.
I attached the link in jira.